lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -4315,14 +4338,17 @@ static void fillCanonicalGradient(SILFunction &canGrad,
   if (primalConv.getResults().back().isFormalDirect())
     directResults.push_back(primalResults.back());
   // Add the adjoint's results to the direct return list.
+  // If the adjoint returns one result, it's the subset we're looking for.
   if (adjointConv.getNumDirectSILResults() == 1)
     directResults.push_back(adjApply);
   else {
     auto tupleSILTy = adjApply->getType();
     for (unsigned i : range(adjointConv.getNumDirectSILResults())) {
-      auto val = builder.createTupleExtract(loc, adjApply, i,
-                                            tupleSILTy.getTupleElementType(i));
-      directResults.push_back(val);
+      if (config.indices.isWrtParameter(i)) {
```

Comment:
This won't work when indirect results are present. Here's an example:
```swift
@differentiable(reverse, adjoint: dfoo)
func foo<T : Differentiable>(x: Float, y: T, z: Float) -> Float

func dfoo<T : Differentiable>(x: Float, y: T, z: Float, originalResult: Float, seed: Float) -> (Float, T, Float) {
    return (..., ..., ...)
}
```

The SIL for `dfoo` is the following:
```swift
sil @dfoo : <T> (Float, @in T, Float, Float, Float) -> (Float, @out T, Float) {
bb0(%0 : $*T, %1 : $Float, %2 : $*T, %3 : $Float, %4 : $Float, %5 : $Float):
//  ^ The T result is indirectly returned
  %ret = tuple (..., ...) : $(Float, Float)
  return %ret : $(Float, Float) // <- this has no T here
}
```

Note that the `T` parameter here can only be indirectly passed. At the call site, the result tuple contains everything but indirect results. We need to map this index onto the direct result index before extracting the result. I can tackle this - if you could file a bug and leave a TODO here, that'd be great.

Reply:
This seems like what you're looking for:
```
// TODO: unbreak this one too
@differentiable(reverse, wrt: (.0, .1), adjoint: dx_T)
func x_T<T>(_ x: Float, _ y: T) -> Float {
  if x > 1000 {
    return x
  }
  return x
}
func dx_T<T>(
    _ x: Float, _ y: T, primal: Float, seed: Float) -> (Float, T) {
  return (x, y)
}
SupersetAdjointTests.test("IndirectResults") {
  expectEqual(3, #gradient(x_T, wrt: .0)(2, Float(3)))
}
```
but it currently throws the "opaque function" error rather than reaching the issue you're describing because `findReferenceToVisibleFunction` fails (all I know is that it takes the second `return nullptr` path and not the first; everything else I try to look at crashes LLDB).

Reply:
Yeah, could you file a bug for “handling indirect results when extracting a subset of primitive adjoint results”?